airbyte-integrations/connectors/source-github/source_github/source.py

```diff
@@ -50,43 +49,57 @@
     WorkflowRuns,
     Workflows,
 )
+from .utils import read_full_refresh
 
 TOKEN_SEPARATOR = ","
 DEFAULT_PAGE_SIZE_FOR_LARGE_STREAM = 10
-# To scan all the repos within orgnaization, organization name could be
-# specified by using asteriks i.e. "airbytehq/*"
-ORGANIZATION_PATTERN = re.compile("^.*/\\*$")
 
 
 class SourceGithub(AbstractSource):
     @staticmethod
-    def _generate_repositories(config: Mapping[str, Any], authenticator: MultipleTokenAuthenticator) -> Tuple[List[str], List[str]]:
+    def _get_org_repositories(config: Mapping[str, Any], authenticator: MultipleTokenAuthenticator) -> Tuple[List[str], List[str]]:
         """
-        Parse repositories config line and produce two lists of repositories.
+        Parse config.repository and produce two lists: organizations, repositories.
         Args:
             config (dict): Dict representing connector's config
             authenticator(MultipleTokenAuthenticator): authenticator object
-        Returns:
-            Tuple[List[str], List[str]]: Tuple of two lists: first representing
-            repositories directly mentioned in config and second is
-            organization repositories from orgs/{org}/repos request.
         """
-        repositories = list(filter(None, config["repository"].split(" ")))
-
-        if not repositories:
+        config_repositories = set(filter(None, config["repository"].split(" ")))
+        if not config_repositories:
             raise Exception("Field `repository` required to be provided for connect to Github API")
 
-        repositories_list: set = {repo for repo in repositories if not ORGANIZATION_PATTERN.match(repo)}
-        organizations = [org.split("/")[0] for org in repositories if org not in repositories_list]
-        organisation_repos = set()
-        if organizations:
-            repos = Repositories(authenticator=authenticator, organizations=organizations)
-            for stream in repos.stream_slices(sync_mode=SyncMode.full_refresh):
-                organisation_repos = organisation_repos.union(
-                    {r["full_name"] for r in repos.read_records(sync_mode=SyncMode.full_refresh, stream_slice=stream)}
-                )
+        repositories = set()
+        organizations = set()
+        unchecked_repos = set()
+        unchecked_orgs = set()
+
+        for org_repos in config_repositories:
+            org, _, repos = org_repos.partition("/")
+            if repos == "*":
+                unchecked_orgs.add(org)
+            else:
+                unchecked_repos.add(org_repos)
+
+        if unchecked_orgs:
+            stream = Repositories(authenticator=authenticator, organizations=unchecked_orgs)
+            for record in read_full_refresh(stream):
+                repositories.add(record["full_name"])
+                organizations.add(record["organization"])
+
+        unchecked_repos = unchecked_repos - repositories
```

Comment:
Is this used so that we don't get duplicates if a user specifies an organization that owns a repository, and the user also specifies that repository individually?

Reply:
yes, for example such config:
```
config["repository"] = "airbytehq/* airbytehq/integration-test"
```

it does not make sense to validate `airbytehq/integration-test` individually .
Because we already validated all repositories of `airbytehq` organization and we already checked `airbytehq/integration-test` too